File: C++/Medium/Accepted/time-needed-to-inform-all-employees_2.cpp
class Solution {
public:
    int numOfMinutes(int n, int headID, vector<int>& manager, vector<int>& informTime) {
        
        vector<vector<int>> subordinates(n);
        for (int i = 0; i < n; i++) {
            if (manager[i] != -1) {
                subordinates[manager[i]].push_back(i);
            }
        }
        return dfs(headID, subordinates, informTime);
    }
    
    int dfs(int id, vector<vector<int>>& subordinates, vector<int>& informTime) {
        int maxTime = 0;
        for (int subId: subordinates[id]) {
            maxTime = max(maxTime, dfs(subId, subordinates, informTime));
        }
        return maxTime + informTime[id];
    }
};

File: Java/Medium/Accepted/time-needed-to-inform-all-employees_2.java
class Solution {
    public int numOfMinutes(int n, int headID, int[] manager, int[] informTime) {
        
        Map<Integer, List<Integer>> map = new HashMap<>();
        for (int i = 0; i < manager.length; i++) {
            if (manager[i] == -1) continue;
            List<Integer> list = map.getOrDefault(manager[i], new ArrayList<>());
            list.add(i);
            map.put(manager[i], list);
        }
        return dfs(map, headID, informTime);
    }
    
    private int dfs(Map<Integer, List<Integer>> map, int cur, int[] informTime) {
        if (!map.containsKey(cur)) return 0;
        int max = 0;
        for (int next : map.get(cur)) {
            max = Math.max(max, dfs(map, next, informTime));
        }
        return max + informTime[cur];
    }
}